core/arch/instr_shared.c

```diff
@@ -3333,7 +3333,14 @@ instr_is_reg_spill_or_restore_ex(void *drcontext, instr_t *instr, bool DR_only,
         reg = &myreg;
     if (instr_check_tls_spill_restore(instr, spill, reg, &check_disp)) {
         int offs = reg_spill_tls_offs(*reg);
-        if (!DR_only || (offs != -1 && check_disp == os_tls_offset((ushort)offs))) {
+        if (!DR_only ||
+            (offs != -1 &&
+             /* Mangling may choose to spill registers to a not natural tls offset,
+              * e.g. rip-rel mangling will, if rax is used by the instruction. We
+              * allow for this here and still recognize the DR spill.
+              */
+             (instr_is_our_mangling(instr) ||
```

Comment:
This only works for a mangled instruction list: i.e., only during bb building (after a client's events) or bb recreation (also after a client's events).   Thus this call will never return true for a client.    Plus, this is-spill function is used for more than instrlists: some callers use it when decoding raw instructions, such as drreg and also I think DR may use it that way when it decodes from the cache in some cases?  So this does not seem a complete solution.

Reply:
>> This only works for a mangled instruction list:
But that was the purpose. Only DR's mangling (I think only rip-rel) can add such a spill. Are there other cases (maybe I should add DR_only &&) where we want to recognize such a spill as well?

Reply:
Won't drreg do the wrong thing if it doesn't recognize this as a DR spill?  Similarly for other cases: though I just looked at decode_fragment() and it only uses instr_is_tls_xcx_spill().

Reply:
That's true. I think we can test for TLS_REG0/1/2/3_SLOT, see patch.